### ISSUE TITLE
Split docker-clean for BuildKit cache

### ISSUE THREAD

**Issue by kohtala** [debuerreotype/docker-debian-artifacts#126]
Apt maintains two directories Dir::State "var/lib/apt" and Dir::Cache "var/cache/apt" that could benefit from BuildKit `RUN --mount=type=cache`.

Use of the cache requires removal of `/etc/apt/apt.conf.d/docker-clean` to remove definitions of DPkg::Post-Invoke and APT::Update::Post-Invoke.

Some builds have additional sources in `/etc/apt/sources.list.d/`, resulting in different `/var/lib/apt/lists/`. But regardless of that difference in lists, it is useful to share the `/var/cache/apt/archives` cache of .deb downloads. However since Dir::Cache::srcpkgcache "" and Dir::Cache::pkgcache "" were in the same removed `/etc/apt/apt.conf.d/docker-clean`, these caches get updated needlessly in `/var/cache/apt/`.

So it would be useful to have Dir::Cache::srcpkgcache "" and Dir::Cache::pkgcache "" in a separate `docker-no-pkgcache`, for example.

Since apt is not good on concurrent access to the caches, the caches need to be used with `sharing=locked`. Therefore this is only a minor issue.

**Comment by tianon** [comment]
Interesting!  Given that this is a somewhat advanced edge use case, I'm inclined to suggest something like the following instead of just deleting the file outright:

```dockerfile
RUN sed -ri -e '/pkgcache/d' /etc/apt/apt.conf.d/docker-clean
```

**Comment by kohtala** [comment]
Yes, something like that occurred to me, but I did not think further because [the documentation](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md#example-cache-apt-packages) already has established the removal of this file. I could have been too pessimistic, because it has such an obvious place to spread best practice and they surely would accept a pull request.

To think of it, the added `Binary::apt::APT::Keep-Downloaded-Packages "true"` could perhaps be added in the same file. Have the sed command change to something like

```Dockerfile
RUN sed -i -e '/#BuildKitRm/,/#BuildKitRm/d' -e 's/#BuildKit: *//' /etc/apt/apt.conf.d/docker-clean
```

would convert `docker-clean` like this

```
#BuildKitRm
DPkg::Post-Invoke { "rm -f /var/cache/apt/archives/*.deb /var/cache/apt/archives/partial/*.deb /var/cache/apt/*.bin || true"; };
APT::Update::Post-Invoke { "rm -f /var/cache/apt/archives/*.deb /var/cache/apt/archives/partial/*.deb /var/cache/apt/*.bin || true"; };
#BuildKitRm

Dir::Cache::pkgcache "";
Dir::Cache::srcpkgcache "";

#BuildKit: Binary::apt::APT::Keep-Downloaded-Packages "true";
```

to this for BuildKit in one command

```
Dir::Cache::pkgcache "";
Dir::Cache::srcpkgcache "";

Binary::apt::APT::Keep-Downloaded-Packages "true";
```

And it could nicely accommodate further changes to `docker-clean` in base images.

**Comment by ecki** [comment]
You mean this doc, right? (It moved)

> https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#example-cache-apt-packages

Btw I wonder if instead of modifying the clean script specifying (-o/ENV - not sure how to skip the post invoke) options to dpkg/apt for the specific RUN is better since it only makes sense in the RUN context during the cache mount (plus for example tmpfs for /var/log…)

Maybe have a envvar and having a `test -n $DEBIAN_DOCKER_KEEPCACHE ||` added to the clean script invoke hooks?

**Comment by kohtala** [comment]
> You mean this doc, right? (It moved)
> 
> > https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#example-cache-apt-packages

Yes.

> Btw I wonder if instead of modifying the clean script specifying (-o/ENV - not sure how to skip the post invoke) options to dpkg/apt for the specific RUN is better since it only makes sense in the RUN context during the cache mount (plus for example tmpfs for /var/log…)

Yes. You have a point.

> Maybe have a envvar and having a `test -n $DEBIAN_DOCKER_KEEPCACHE ||` added to the clean script invoke hooks?

I'd put `test -z $APT_KEEP_ARCHIVES &&`. The `test -n` is true if the parameter is missing due to undefined environment variable.

With that change `APT_KEEP_ARCHIVES=1 apt -o APT::Keep-Downloaded-Packages=true install` would keep the deb files.
